### PR TITLE
Allow callinfo live on-stack

### PR DESCRIPTION
**Pull request by shyouhei**
`struct rb_callinfo`s are created on-the-fly, which results in increasing GC pressure.  However they include no references to other objects, and those on-the-fly CIs tend to be short lived.  Why not skip allocation of them.  By allowing CI/CC to be on stack, some functions could be reduced in generated binary.

Main purpose of this pull request is the binary size.  No speedup is observed for optcarrot.  However because it modifies `rb_vm_call0`, some micro benchmarks cerebrate very small improvements.

```
Calculating -------------------------------------
                         master        ours
 enum_lazy_grep_v_20      6.481       6.600 i/s -       1.000 times in 0.154301s 0.151505s
 enum_lazy_grep_v_50      4.846       4.953 i/s -       1.000 times in 0.206348s 0.201884s
enum_lazy_grep_v_100      3.430       3.733 i/s -       1.000 times in 0.291519s 0.267860s
   enum_lazy_uniq_20      5.589       5.818 i/s -       1.000 times in 0.178924s 0.171895s
   enum_lazy_uniq_50      4.042       4.152 i/s -       1.000 times in 0.247420s 0.240838s
  enum_lazy_uniq_100      2.772       2.862 i/s -       1.000 times in 0.360692s 0.349366s

Comparison:
              enum_lazy_grep_v_20
                ours:         6.6 i/s
              master:         6.5 i/s - 1.02x  slower

              enum_lazy_grep_v_50
                ours:         5.0 i/s
              master:         4.8 i/s - 1.02x  slower

             enum_lazy_grep_v_100
                ours:         3.7 i/s
              master:         3.4 i/s - 1.09x  slower

                enum_lazy_uniq_20
                ours:         5.8 i/s
              master:         5.6 i/s - 1.04x  slower

                enum_lazy_uniq_50
                ours:         4.2 i/s
              master:         4.0 i/s - 1.03x  slower

               enum_lazy_uniq_100
                ours:         2.9 i/s
              master:         2.8 i/s - 1.03x  slower

```

### CODE REVIEW

File: vm_eval.c [hunk 1]
```diff
@@ -45,12 +45,20 @@ static VALUE vm_call0_body(rb_execution_context_t* ec, struct rb_calling_info *c
 MJIT_FUNC_EXPORTED VALUE
 rb_vm_call0(rb_execution_context_t *ec, VALUE recv, ID id, int argc, const VALUE *argv, const rb_callable_method_entry_t *me, int kw_splat)
 {
-    struct rb_calling_info calling = { Qundef, recv, argc, kw_splat, };
-    const struct rb_callinfo *ci = vm_ci_new_runtime(id, kw_splat ? VM_CALL_KW_SPLAT : 0, argc, NULL);
-    struct rb_callcache cc_body;
-    const struct rb_callcache *cc = vm_cc_fill(&cc_body, 0, me, vm_call_general);
-    struct rb_call_data cd = { ci, cc, };
-    return vm_call0_body(ec, &calling, &cd, argv);
+    return vm_call0_body(
```
Comment (ko1) [inline vm_eval.c:48]: just curious. why do you use compound literal? for me the previous version is easy to read.

File: iseq.c [hunk 2]
```diff
@@ -325,9 +325,13 @@ rb_iseq_mark(const rb_iseq_t *iseq)
         if (body->call_data) {
```
Comment (ko1) [inline iseq.c:325]: from commit log:
> "to increase GC pressure"

to reduce?

Comment (ko1) [reply iseq.c:325]: ah, does it means "on the fly CI creation will increase the pressure"?

**Comment by ko1** [comment]
I understand the idea is packed or allocated CIs are not needed for some cases and they can be replaced with fake (no gc managed) IMEMO objects on a machine stack. It seems reasonable. 

On concern: I designed it under the assumption that most of case packed CIs are used, and the code will expect packed CIs are given and CPU branch prediction can miss with IMEMO CIs. Not critical, but I want to note it to revisit it.

File: iseq.c (hunk 2 above)
Comment (shyouhei) [reply iseq.c:325]: Yes.
Screenshot from 2020-06-03 18-04-28

File: vm_eval.c (hunk 1 above)
Comment (shyouhei) [reply vm_eval.c:48]: Not the compound literal (`&(struct rb_call_data)`) part, but I prefer designated initializers (`.ci = ..., .cc = ..., `) part.  They prevent casual field order confusions.

File: vm_eval.c [hunk 3]
```diff
@@ -45,12 +45,20 @@ static VALUE vm_call0_body(rb_execution_context_t* ec, struct rb_calling_info *c
...
+            .recv = recv,
+            .argc = argc,
+            .kw_splat = kw_splat,
+        },
+        &(struct rb_call_data) {
+            .ci = &VM_CI_ON_STACK(id, kw_splat ? VM_CALL_KW_SPLAT : 0, argc, NULL),
+            .cc = &VM_CC_ON_STACK(Qfalse, vm_call_general, { 0 }, me),
+        },
+        argv
+    );
```
Comment (shyouhei) [inline vm_eval.c:61]: @ko1 This is acceptable to me.  Thoughts?
```suggestion
    struct calling_info calling = {
        .block_handler = VM_BLOCK_HANDLER_NONE,
        .recv = recv,
        .argc = argc,
        .kw_splat = kw_splat,
    };
    struct rb_call_data cd = {
        .ci = &VM_CI_ON_STACK(id, kw_splat ? VM_CALL_KW_SPLAT : 0, argc, NULL),
        .cc = &VM_CC_ON_STACK(Qfalse, vm_call_general, { 0 }, me),
    };
    
    return vm_call0_body(ec, &calling, &cd, argv);
```

File: iseq.c (hunk 2 above)
Comment (ko1) [reply iseq.c:325]: The purpose is not increasing GC pressure...

File: vm_eval.c (hunk 1 above)
Comment (ko1) [reply vm_eval.c:48]: I understand the Designated Initializers are useful.

File: iseq.c (hunk 2 above)
Comment (shyouhei) [reply iseq.c:325]: Sorry about that.  Let me rephrase in a moderate tone (will amend the commit message).

File: vm_eval.c (hunk 3 above)
Comment (ko1) [reply vm_eval.c:61]: I like this style. But as I mention it as "curious", I want to ask the preference.

Comment (shyouhei) [reply vm_eval.c]: Preference for compound literals...

Well I think "having no variable" is a good property.  That way a struct is implicitly immutable, because there is no way to reuse it.  If you assign it to a variable, you have to manually make sure it won't be clobbered later.

Also, "having no need to think about variable name" is easier to me.  That's not a big pain, but a pain.

Comment (ko1) [reply vm_eval.c]: Agreed for the reasons.
I feel it is difficult because the deep nesting and difficult to understand the structure.
I'm not sure which should be prioritize...

**Comment by shyouhei** [comment]
I don't think the failing CIs are my faults.  I think this branch is ready to be merged.
